### PR TITLE
[RFC] handle runtime shutdowns while tasks are being spawned

### PR DESCRIPTION
**Pull request by sunshowers**
If a runtime has been marked as shutting down before `spawn_blocking` is
called, or before the blocking task is first scheduled, the `JoinHandle` will
return a cancellation error. Currently, this panics, which tends to manifest as
test failures (e.g. https://github.com/oxidecomputer/omicron/issues/6505)

Instead, add a `RunError` wrapper that captures the cancellation situation.
Callers might choose to bubble this error up or to squelch it -- unfortunately,
we can't make this decision.

I haven't written tests yet but I should be able to if this looks reasonble.

The API surface here is really unfortunate, but I can't see a better way to do
this than to propagate errors up the stack.

### CODE REVIEW

**Comment by sunshowers** [comment]
I really don't know how to feel about this! It really sucks and adds a significant extra burden to callers, but I think it might be necessary.

* If Diesel were async and we weren't using tasks, this wouldn't be a problem because the calling code would also get shut down as part of the same task.
* If Tokio had a way to "propagate cancellations across tasks", e.g. by panicking the current task with a special type rather than returning a value, this also wouldn't be a problem. (Though that might still not be good enough at squelching errors properly.)
* Unfortunately we must return a value, so we have to add an enum variant to return.

**Comment by sunshowers** [comment]
@hawkw wonder if you have thoughts about this. Feels unfortunate how janky this is

File: src/error.rs [hunk 1]
```diff
@@ -44,6 +56,40 @@ impl<T> OptionalExtension<T> for Result<T, ConnectionError> {
...
+
+        self_as_query_result
+            .optional()
+            .map_err(ConnectionError::Query)
+    }
+}
+
+/// Errors encountered while running a function on a connection pool.
+#[derive(Error, Debug, Clone, Copy, Eq, PartialEq)]
+pub enum RunError<E> {
```
Comment (smklein) [inline src/error.rs]: Does this actually need to be generic? Could we just have `E` be `DieselError`?

I'm aware that (before this PR) there were spots where we used `E: From<DieselError>` -- seems like those spots could become `E: From<RunError>` instead, but without making `RunError` itself be generic.

File: src/async_traits.rs [hunk 2]
```diff
@@ -371,53 +390,86 @@ where
...
-                    Err(err) => Err(err),
+                    Err(RunError::User(err)) => Err(err),
+                    Err(RunError::RuntimeShutdown) => Err(RunError::RuntimeShutdown.into()),
                 }
             }
         }
     }
 }
 
+fn handle_spawn_blocking_error<T, E>(
```
Comment (smklein) [inline src/async_traits.rs]: If we're going to be propagating errors out during runtime shutdown, this seems reasonable to me (modulo the comment on whether we can collapse one layer of generics, see my other comment).

However, I'm still a little confused here - before this PR, we had a call-stack of:
- User code
- ... calls async_bb8_diesel API
- ... which tries to spawn a task and fails, leading to a panic

Now, we'll have:
- User code
- ... calls async_bb8_diesel API
- ... which tries to spawn a task and fails, returning an error

But what do we expect the caller to do? In the shutdown case, won't the caller most likely panic anyway?

Put another way -- what useful thing do we expect the caller to do **other** than panic here?

File: src/async_traits.rs [hunk 3]
```diff
@@ -371,53 +390,86 @@ where
...
+                match conn
+                    .run_with_shared_connection(|conn| {
+                        Conn::TransactionManager::commit_transaction(conn)
+                            .map_err(|err| E::from(RunError::User(err)))
+                    })
+                    .await
+                {
+                    Ok(()) => Ok(value),
+                    // XXX: we should try to roll this back
+                    Err(RunError::User(err)) => Err(err),
```
Comment (smklein) [inline src/async_traits.rs]: I believe this is automatically done by the `commit_transaction` method in the case where we return an error.

See: https://docs.diesel.rs/master/src/diesel/connection/transaction_manager.rs.html#451

It *looks* l like `get_transaction_state` could return an error and stop us from propagating an error here, but I think that only happens if the TransactionManager has already been marked "broken, do not use" by a different call. So, ignoring that error-return-path, [if we fail to commit, we automatically perform rollback](https://docs.diesel.rs/master/src/diesel/connection/transaction_manager.rs.html#487-508)

TL;DR: I think this comment is unnecessary, we already do try to roll back.

File: src/async_traits.rs (hunk 2 above)
Comment (smklein) [reply src/async_traits.rs]: Also, to be clear, I'm asking this question not because I believe propagating the error is wrong, but because I fear I may have a gap in my understanding of "why this will make the https://github.com/oxidecomputer/omicron/issues/6505 issues less bad".

Comment (sunshowers) [reply src/async_traits.rs]: Ah yeah that's what I was talking about in my comments — the caller would have to learn to squelch the error.

Basically there is some point in the call stack at which the return value would be (), or some other type with a sensible default value. At that point, the caller would have to examine the error and immediately return the default value.

One alternative would be to move the runtime shutdown situation into the Ok case but that feels even weirder.

Definitely not a great situation.

File: src/async_traits.rs (hunk 3 above)
Comment (sunshowers) [reply src/async_traits.rs]: Thanks!

File: src/async_traits.rs (hunk 2 above)
Comment (sunshowers) [reply src/async_traits.rs]: I think a useful next step would be to try patching this PR into Omicron and seeing how to manage this situation.

File: src/error.rs (hunk 1 above)
Comment (sunshowers) [reply src/error.rs]: I'm not sure! I'll try it out. I was trying to minimize changes in the PR but not having a generic here would certainly simplify things.

File: src/async_traits.rs [hunk 4]
```diff
@@ -371,53 +390,86 @@ where
...
+                // spawn_blocking was called.
+                Err(RunError::RuntimeShutdown)
+            } else if err.is_panic() {
+                // Propagate panics.
+                std::panic::panic_any(err.into_panic());
+            } else {
+                // Not possible to reach this as of Tokio 1.40, but maybe in
+                // future versions.
+                panic!("unexpected JoinError: {:?}", err);
+            }
```
Comment (hawkw) [inline src/async_traits.rs]: very minor style suggestion, take it or leave it: i might roll the if conditions into the match, personally:
```suggestion
        // The only way a spawn_blocking task can be marked cancelled
        // is if the runtime started shutting down _before_
        // spawn_blocking was called.
        Err(err) if err.is_cancelled() => Err(RunError::RuntimeShutdown),
        // Propagate panics
        Err(err) if err.is_panic() => std::panic::panic_any(err.into_panic()),
        // Not possible to reach this as of Tokio 1.40, but maybe in
        // future versions.
        Err(err) => panic!("unexpected JoinError: {:?}", err),
```

File: src/async_traits.rs [hunk 5]
```diff
@@ -371,53 +390,86 @@ where
...
+                // is if the runtime started shutting down _before_
+                // spawn_blocking was called.
+                Err(RunError::RuntimeShutdown)
+            } else if err.is_panic() {
+                // Propagate panics.
+                std::panic::panic_any(err.into_panic());
+            } else {
+                // Not possible to reach this as of Tokio 1.40, but maybe in
+                // future versions.
+                panic!("unexpected JoinError: {:?}", err);
```
Comment (hawkw) [inline src/async_traits.rs]: also tremendously nitpicky; i _might_ consider `unreachable!()` here, and i might also reword the panic message to make it explicit that this is likely a result of a new Tokio version adding new join error variants?

File: src/error.rs [hunk 6]
```diff
@@ -44,6 +56,40 @@ impl<T> OptionalExtension<T> for Result<T, ConnectionError> {
...
+    #[error("error in user code")]
+    User(#[from] E),
```
Comment (hawkw) [inline src/error.rs]: IMO, this should probably be `#[error(transparent)]` --- I don't think there's much value in replacing the formatted representation of the underlying error with the very opaque "error in user code", and only displaying the underlying error if the user displays the cause chain.
```suggestion
    #[error(transparent)]
    User(#[from] E),
```

File: src/async_traits.rs [hunk 7]
```diff
@@ -262,20 +263,26 @@ where
                         //
                         // We're still in the transaction, but we at least
                         // tried to ROLLBACK to our savepoint.
-                        if !retryable_error(&err) || !retry().await {
+                        let retried = match &err {
+                            RunError::User(err) => retryable_error(err) && retry().await,
+                            RunError::RuntimeShutdown => false,
+                        };
```
Comment (hawkw) [inline src/async_traits.rs]: wondered briefly if `retryable_error` ought to just take a `RunError`, but it looks like it's only called in two places, and one of them is a place where we've already extracted the user error, so...nevermind.

**Review by hawkw — COMMENTED** [review]
So. I agree that this is a bit unfortunate, but I also agree that the only better solution is just "have upstream code promise to never shut down the runtime", which I don't know if we can reasonably do. I do agree with https://github.com/oxidecomputer/async-bb8-diesel/pull/77#discussion_r1799955750 that we should probably see what callers in Omicron would actually do with the new error variant before committing to this change. 

Beyond that, I had some very small suggestions, but none of them are particularly important to me.

File: src/async_traits.rs (hunk 3 above)
Comment (sunshowers) [reply src/async_traits.rs]: Removed the comment.

File: src/error.rs (hunk 1 above)
Comment (sunshowers) [reply src/error.rs]: Removing the type parameter is possible as long as we change the `run` callback to always return a `DieselError`. Another option is to introduce a separate error type just for that function.

File: src/async_traits.rs (hunk 5 above)
Comment (sunshowers) [reply src/async_traits.rs]: Changed the panic message. I deliberately didn't use unreachable because I felt that it was potentially reachable with future Tokio versions, I think of unreachable as handling cases where you can prove that a line of code won't be hit (e.g. you checked for a value earlier in a function, then panicked on seeing that value again). wdyt?

File: src/error.rs (hunk 6 above)
Comment (sunshowers) [reply src/error.rs]: Done, I swore I had changed this before putting it up...

File: src/async_traits.rs (hunk 7 above)
Comment (sunshowers) [reply src/async_traits.rs]: Yeah

**Comment by sunshowers** [comment]
Tried integrating this into omicron: https://github.com/oxidecomputer/omicron/pull/6875. It does appear to be possible but it's a lot of work for marginal gain.

https://github.com/oxidecomputer/qorb/pull/45 should work well enough, so closing this.
